### PR TITLE
feat: add BitVec.intMin

### PR DESCRIPTION
**Pull request by tobiasgrosser**
We keep the library interface for `intMin` minimal and make it an `abbrev` for `twoPow`. To ensure a uniform library interface, we also turn `intMax` into an abbrev for `allOnes`.

To improve proof automation, we introduce a simp-set `boolToPropSimps` plus corresponding lemmas that turn our `intMin_toLsb` proof into a simple `simp; omega` proof. This simp set is not yet active, but should become so after a stage-0 update.

This PR also pulls in some mathlib theorems on testBit and Nat and establishes facts about `2 ^ w` that are needed here and which are generally useful for bitvector reasoning.

The following theorem is not generalized to arbitrary `x` instead of `2`, as this would require a condition to be added for `x > 1` which would have to be passed to simp each time this theorem should fire:

```lean
protected theorem two_pow_pred_mod_two_pow (h : 0 < w) :
    2 ^ (w - 1) % 2 ^ w = 2 ^ (w - 1)
```

Mathlib reports only errors for theorems that this PR moves into lean.

```
error: ././././Mathlib/Data/Nat/Bitwise.lean:250:8: 'Nat.testBit_two_pow_self' has already been declared
error: ././././Mathlib/Data/Nat/Bitwise.lean:254:8: 'Nat.testBit_two_pow_of_ne' has already been declared
error: ././././Mathlib/Data/Nat/Bitwise.lean:265:8: 'Nat.testBit_two_pow' has already been declared
```

### CODE REVIEW

**Comment by tobiasgrosser** [comment]
Draft, as I am still polishing the order of the various theorems within the files. Any feedback is obviously already welcome.

File: src/Init/Data/BitVec/Lemmas.lean [hunk 1]
```diff
@@ -1449,6 +1449,32 @@ protected theorem lt_of_le_ne (x y : BitVec n) (h1 : x <= y) (h2 : ¬ x = y) : x
   simp
   exact Nat.lt_of_le_of_ne
 
+/-! ### intMin -/
+
+/-- The bitvector of width `w` that has the smallest value when interpreted as an integer. -/
+def intMin (w : Nat) : BitVec w := BitVec.ofNat w (2^(w - 1))
+
+theorem getLsb_intMin (w : Nat) : (intMin w).getLsb i = decide (i + 1 = w) := by
+  simp [intMin, BitVec.getLsb]
```
Comment (hargoniX) [inline src/Init/Data/BitVec/Lemmas.lean:1458]: Non terminal simp

File: src/Init/Data/BitVec/Lemmas.lean [hunk 2]
```diff
@@ -1449,6 +1449,32 @@ protected theorem lt_of_le_ne (x y : BitVec n) (h1 : x <= y) (h2 : ¬ x = y) : x
...
+  by_cases h : i + 1 = w
```
Comment (hargoniX) [inline src/Init/Data/BitVec/Lemmas.lean:1459]: There is a lemma for this kind of case distinction: [Nat.lt_trichotomy](https://leanprover-community.github.io/mathlib4_docs/Init/Data/Nat/Basic.html#Nat.lt_trichotomy)

File: src/Init/Data/BitVec/Lemmas.lean [hunk 3]
```diff
@@ -1449,6 +1449,46 @@ protected theorem lt_of_le_ne (x y : BitVec n) (h1 : x <= y) (h2 : ¬ x = y) : x
...
+
+theorem getLsb_intMin' (w : Nat) : (intMin w).getLsb i = decide (i + 1 = w) := by
+  simp only [getLsb, intMin, toNat_ofNat, Nat.testBit_mod_two_pow]
+  by_cases h : i + 1 = w
+  · subst h
+    simp
+  · by_cases i + 1 < w
+    · simp [h, @Nat.testBit_two_pow_of_ne (w-1) i (by omega)]
+    · simp only [h, decide_False]
+      simp only [Bool.and_eq_false_imp, decide_eq_true_eq] ;
```
Comment (AtticusKuhn) [inline src/Init/Data/BitVec/Lemmas.lean:1465]: could these two simps be combined into one simp?

**Review by AtticusKuhn — APPROVED** [review]
looks good to me.

File: src/Init/Data/BitVec/Lemmas.lean [hunk 4]
```diff
@@ -1449,6 +1449,46 @@ protected theorem lt_of_le_ne (x y : BitVec n) (h1 : x <= y) (h2 : ¬ x = y) : x
...
+theorem toNat_intMin : (intMin w).toNat = 2^(w - 1) % 2^w := by
+  simp [intMin]
+
+@[simp]
+theorem neg_intMin {w : Nat} : -intMin w = intMin w := by
+  by_cases h : 0 < w
+  · simp [bv_toNat, h]
+  · simp only [Nat.not_lt, Nat.le_zero_eq] at h
+    subst h
+    simp [intMin]
```
Comment (AtticusKuhn) [inline src/Init/Data/BitVec/Lemmas.lean:1490]: How is "subst" different from "rw" or "simp"? (I tried to look up the documentation, did not find much). If "subst" just provides substitution, then maybe this proof could be shortened to 
```lean
simp [h, intMin]
```

Comment (hargoniX) [reply src/Init/Data/BitVec/Lemmas.lean:1489]: `subst` does indeed "just provide substitution". However in certain DTT hell scenarios `subst` is often able to perform substitutions along an equality while `simp` and `rw` cannot due to the way that they work. Whether that is the case here I do not know.

Comment (tobiasgrosser) [reply src/Init/Data/BitVec/Lemmas.lean:1489]: I tried, but here the order seems to be relevant.

**Comment by tobiasgrosser** [comment]
This is now ready-for-review from my side.

**Comment by tobiasgrosser** [comment]
awaiting-review

File: src/Init/Data/Nat/Bitwise/Lemmas.lean [hunk 5]
```diff
@@ -324,6 +324,22 @@ theorem testBit_one_eq_true_iff_self_eq_zero {i : Nat} :
...
+@[simp]
+theorem testBit_two_pow_of_ne {n m : Nat} (hm : n ≠ m) : Nat.testBit (2 ^ n) m = false := by
+  rw [testBit, shiftRight_eq_div_pow]
+  cases Nat.lt_or_lt_of_ne hm
+  · rw [div_eq_of_lt (Nat.pow_lt_pow_of_lt (by omega) (by omega))]
+    simp
+  · rw [Nat.pow_div _ Nat.two_pos,
+       ← Nat.sub_add_cancel (succ_le_of_lt <| Nat.sub_pos_of_lt (by omega))]
+    simp [Nat.pow_succ, and_one_is_mod, mul_mod_left]
+    omega
```
Comment (semorrison) [inline src/Init/Data/Nat/Bitwise/Lemmas.lean:341]: Could you please derive these from a (non-`@[simp]`) `testBit_two_pow`, with an if statement for the RHS?

File: src/Init/Data/Nat/Lemmas.lean [hunk 6]
```diff
@@ -717,6 +717,32 @@ protected theorem pow_lt_pow_iff_right {a n m : Nat} (h : 1 < a) :
...
+    x ^ (w - 1) * x = x ^ w := by
+  simp [← Nat.pow_succ, succ_eq_add_one, Nat.sub_add_cancel h]
+
+protected theorem pow_pred_lt_pow {x w : Nat} (h1 : 1 < x) (h2 : 0 < w) :
+    x ^ (w - 1) < x ^ w :=
+  @Nat.pow_lt_pow_of_lt x (w - 1) w  h1 (by omega)
+
+@[simp]
+protected theorem two_pow_pred_add_two_pow_pred (h : 0 < w) :
+    2 ^ (w - 1) + 2 ^ (w - 1) = 2 ^ w:= by
```
Comment (semorrison) [inline src/Init/Data/Nat/Lemmas.lean:731]:
```suggestion
    2 ^ (w - 1) + 2 ^ (w - 1) = 2 ^ w := by
```

File: src/Init/Data/Nat/Lemmas.lean [hunk 7]
```diff
@@ -717,6 +717,32 @@ protected theorem pow_lt_pow_iff_right {a n m : Nat} (h : 1 < a) :
...
     exact Nat.pow_lt_pow_of_lt h w
 
+@[simp]
+protected theorem pow_pred_mul {x w : Nat} (h : 0 < w) :
+    x ^ (w - 1) * x = x ^ w := by
+  simp [← Nat.pow_succ, succ_eq_add_one, Nat.sub_add_cancel h]
+
+protected theorem pow_pred_lt_pow {x w : Nat} (h1 : 1 < x) (h2 : 0 < w) :
+    x ^ (w - 1) < x ^ w :=
+  @Nat.pow_lt_pow_of_lt x (w - 1) w  h1 (by omega)
```
Comment (semorrison) [inline src/Init/Data/Nat/Lemmas.lean:727]:
```suggestion
protected theorem pow_pred_lt_pow {x w : Nat} (h₁ : 1 < x) (h₂ : 0 < w) :
    x ^ (w - 1) < x ^ w :=
  Nat.pow_lt_pow_of_lt h₁ (by omega)
```

**Comment by semorrison** [comment]
As we are modifying `simp` lemmas for `Nat`, I would like to see Mathlib tested against this. If you rebase this PR onto `nightly-with-mathlib`, you should hopefully get a `builds-mathlib` or `breaks-mathlib` label within an hour.

File: src/Init/Data/BitVec/Lemmas.lean [hunk 8]
```diff
@@ -1446,6 +1446,36 @@ protected theorem lt_of_le_ne (x y : BitVec n) (h1 : x <= y) (h2 : ¬ x = y) : x
...
+  rcases Nat.lt_trichotomy (i + 1) w with lt | rfl | gt
+  case inl =>
+    rw [Nat.testBit_two_pow_of_ne (by omega)]
+    simp
+    omega
+  case inr.inl =>
+    simp
+  case inr.inr =>
+    simp [show ¬ (i + 1 = w) by omega]
+    omega
```
Comment (semorrison) [inline src/Init/Data/BitVec/Lemmas.lean:1465]: I'm not super happy with this. I think `intMin` and `intMax` should be moved below `twoPow`.

Then use `def intMin (w : Nat) : BitVec w := twoPow w (w - 1)`, and `def intMax (w : Nat) : BitVec w := allOnes w`.

In fact both could be abbrevs, or just not get defined?

Every new def adds lots of missing API which is not keeping up.

Comment (semorrison) [reply src/Init/Data/BitVec/Lemmas.lean:1465]: Moreover, the proof of `getLsb_intMin` should be `by simp [intMin, boolToPropSimps]; omega`, for a simp set that we don't have yet that converts statements about `decide` into the corresponding propositional formulas.

File: src/Init/Data/Nat/Bitwise/Lemmas.lean (hunk 5 above)
Comment (tobiasgrosser) [reply src/Init/Data/Nat/Bitwise/Lemmas.lean]: Done.

File: src/Init/Data/BitVec/Lemmas.lean (hunk 8 above)
Comment (tobiasgrosser) [reply src/Init/Data/BitVec/Lemmas.lean]: Done.

**Comment by tobiasgrosser** [comment]
The comments are adressed, except for finding the correct imports for the new simp set. I will await advice from @bollu.

**Comment by tobiasgrosser** [comment]
The simp-set is also defined, so the remaining issue is a build failure in the nix build, which I am not sure how to debug.

File: src/Init/Data/Nat/Bitwise/Lemmas.lean [hunk 9]
```diff
@@ -324,6 +324,28 @@ theorem testBit_one_eq_true_iff_self_eq_zero {i : Nat} :
     Nat.testBit 1 i = true ↔ i = 0 := by
   cases i <;> simp
 
+theorem testBit_two_pow {n m : Nat} : Nat.testBit (2 ^ n) m = if n = m then true else false := by
```
Comment (bollu) [inline src/Init/Data/Nat/Bitwise/Lemmas.lean:327]:
```suggestion
theorem testBit_two_pow {n m : Nat} : Nat.testBit (2 ^ n) m = if n = m then true else false := by
```
The `simp` normal form is `⊢ (2 ^ n).testBit m = decide (n = m)`, and I think this should perhaps even live in `Nat/Bitwise/Lemmas.lean`, but I leave that up to Kim :)

Comment (tobiasgrosser) [reply src/Init/Data/Nat/Bitwise/Lemmas.lean]: Resolved.

File: src/Init/Data/BitVec/Lemmas.lean [hunk 10]
```diff
@@ -1783,4 +1769,38 @@ theorem getLsb_replicate {n w : Nat} (x : BitVec w) :
...
+
+/-! ### intMax -/
+
+/-- The bitvector of width `w` that has the largest value when interpreted as an integer. -/
+abbrev intMax (w : Nat) := allOnes w
+
+theorem getLsb_intMax (w : Nat) : (intMax w).getLsb i = decide (i < w) := by
+  simp [getLsb]
+
+theorem toNat_intMax : (intMax w).toNat = 2 ^ w - 1 := by
```
Comment (bollu) [inline src/Init/Data/BitVec/Lemmas.lean:1800]:
```suggestion
@[simp bv_toNat]
theorem toNat_intMax : (intMax w).toNat = 2 ^ w - 1 := by
```

File: src/Init/Data/BitVec/Lemmas.lean [hunk 11]
```diff
@@ -1783,4 +1769,38 @@ theorem getLsb_replicate {n w : Nat} (x : BitVec w) :
...
+  · simp [bv_toNat, h]
+  · simp only [Nat.not_lt, Nat.le_zero_eq] at h
+    simp [bv_toNat, h]
+
+/-! ### intMax -/
+
+/-- The bitvector of width `w` that has the largest value when interpreted as an integer. -/
+abbrev intMax (w : Nat) := allOnes w
+
+theorem getLsb_intMax (w : Nat) : (intMax w).getLsb i = decide (i < w) := by
```
Comment (bollu) [inline src/Init/Data/BitVec/Lemmas.lean:1797]:
```suggestion
@[simp]
theorem getLsb_intMax (w : Nat) : (intMax w).getLsb i = decide (i < w) := by
```

**Comment by tobiasgrosser** [comment]
It seems that the intMax definition was incorrect. I corrected it, but I haven't polished the proofs yet.

File: src/Init/Data/Nat/Lemmas.lean [hunk 12]
```diff
@@ -717,6 +720,39 @@ protected theorem pow_lt_pow_iff_right {a n m : Nat} (h : 1 < a) :
...
+    x ^ (w - 1) * x = x ^ w := by
+  simp [← Nat.pow_succ, succ_eq_add_one, Nat.sub_add_cancel h]
+
+@[simp]
+protected theorem one_ge_two_pow {w : Nat} :
+   1 ≤ 2 ^ w := by
+  have h₂ := @Nat.pow_le_pow_left 1 2 (by omega) w
+  simp only [Nat.one_pow] at h₂
+  omega
+
```
Comment (tobiasgrosser) [inline src/Init/Data/Nat/Lemmas.lean:734]:
```suggestion
```

File: src/Init/Data/BitVec/Lemmas.lean [hunk 13]
```diff
@@ -1783,4 +1769,60 @@ theorem getLsb_replicate {n w : Nat} (x : BitVec w) :
...
+theorem getLsb_intMax (w : Nat) : (intMax w).getLsb i = decide (i + 1 < w) := by
+  simp only [getLsb, bv_toNat]
+  by_cases h : w = 0
+  · simp [h]
+  · rw [Nat.mod_eq_of_lt (a := 2 ^ (w - 1)) (Nat.pow_pred_lt_pow (by omega) (by omega)),
+      Nat.one_mod_two_pow_eq_one.mpr (by omega), ← Nat.add_sub_assoc (by simp),
+      Nat.sub_add_comm (by simp)]
+    simp only [Nat.add_mod_right, Nat.testBit_mod_two_pow, Nat.testBit_two_pow_sub_one,
+      Bool.and_eq_decide, decide_eq_decide]
+    omega
```
Comment (semorrison) [inline src/Init/Data/BitVec/Lemmas.lean:1826]:
```suggestion
@[simp, bv_toNat]
theorem toNat_intMax : (intMax w).toNat = 2 ^ (w - 1) - 1 := by
  simp only [intMax]
  by_cases h : w = 0
  · simp [h]
  · have h' : 0 < w := by omega
    rw [toNat_sub, toNat_twoPow, ← Nat.sub_add_comm (by simpa [h'] using Nat.one_le_two_pow),
      Nat.add_sub_assoc (by simpa [h'] using Nat.one_le_two_pow),
      Nat.two_pow_pred_mod_two_pow h', ofNat_eq_ofNat, toNat_ofNat, Nat.one_mod_two_pow h',
      Nat.add_mod_left, Nat.mod_eq_of_lt]
    have := Nat.two_pow_pred_lt_two_pow h'
    have := Nat.two_pow_pos w
    omega

@[simp]
theorem getLsb_intMax (w : Nat) : (intMax w).getLsb i = decide (i + 1 < w) := by
  rw [← testBit_toNat, toNat_intMax, Nat.testBit_two_pow_sub_one, decide_eq_decide]
  omega
```

File: src/Init/Data/Nat/Lemmas.lean [hunk 14]
```diff
@@ -717,6 +720,32 @@ protected theorem pow_lt_pow_iff_right {a n m : Nat} (h : 1 < a) :
...
 
+@[simp]
+protected theorem pow_pred_mul {x w : Nat} (h : 0 < w) :
+    x ^ (w - 1) * x = x ^ w := by
+  simp [← Nat.pow_succ, succ_eq_add_one, Nat.sub_add_cancel h]
+
+protected theorem pow_pred_lt_pow {x w : Nat} (h₁ : 1 < x) (h₂ : 0 < w) :
+    x ^ (w - 1) < x ^ w :=
+  Nat.pow_lt_pow_of_lt h₁ (by omega)
+
```
Comment (semorrison) [inline src/Init/Data/Nat/Lemmas.lean:731]:
```suggestion
protected theorem pow_pred_lt_pow {x w : Nat} (h₁ : 1 < x) (h₂ : 0 < w) :
    x ^ (w - 1) < x ^ w :=
  Nat.pow_lt_pow_of_lt h₁ (by omega)

protected theorem two_pow_pred_lt_two_pow {w : Nat} (h : 0 < w) :
    2 ^ (w - 1) < 2 ^ w :=
  Nat.pow_pred_lt_pow (by omega) h

```

**Comment by semorrison** [comment]
Remaining steps to activating `boolToPropSimps`:
* Move `BoolToPropSimps.lean` to `src/Init/Data/`.
* add `import Init.Data.BoolToPropSimps` to `src/Init/Data.lean`
* commit that
* `make -j32 -C build/release update-stage0`
* `git commit -m "chore: update stage0"`
  (it's essential the commit message is exactly that)
* now uncomment the `@[boolToPropSimps]` attributes, and hopefully you can use `simp only [boolToPropSimps]`.

(If you give me write access on your repo I'm happy to do these steps, too.)

**Comment by tobiasgrosser** [comment]
> Remaining steps to activating `boolToPropSimps`:
> 
> * Move `BoolToPropSimps.lean` to `src/Init/Data/`.
> * add `import Init.Data.BoolToPropSimps` to `src/Init/Data.lean`
> * commit that
> * `make -j32 -C build/release update-stage0`
> * `git commit -m "chore: update stage0"`
>   (it's essential the commit message is exactly that)
> * now uncomment the `@[boolToPropSimps]` attributes, and hopefully you can use `simp only [boolToPropSimps]`.
> 
> (If you give me write access on your repo I'm happy to do these steps, too.)

Cool. I gave you commit access. I guess its easier if you push this through.

**Pull request merged**